### PR TITLE
(basic) license compatiblity check

### PR DESCRIPTION
**Pull request by adrianjost**
# About
I'm unsure if I handle everything 100% correct, but I found no ready to use npm package that worked without issues. I have tested:
- legally
- license-checker
- license-compatibility-checker
- node-license-validator
- and some more I don't remember anymore

The current implemenetation should be a good baseline to prevent obvious license conflicts, even if it's not handling every case 100% correctly.

There exist an easy to use commercial version, but it's pricy: https://fossa.com/pricing/.
If you know any better implementation, feel free to comment 😃.

## How it works
1. The Script is extracting the current license from the package.json.
2. Then it checks, which licenses are compatible with the current license using the provided json file.
3. All (production) dependency licenses get extracted and checked if they are in the compatible list.

# License compatiblity

### CODE REVIEW

**Comment by janrenz** [comment]
Nice idea!

**Comment by adrianjost** [comment]
@janrenz I'm working on a npm package version. When finished we could use it in more repos

**Comment by adrianjost** [comment]
@janrenz 
Done, it's now using the [license-ci-checker](https://www.npmjs.com/package/license-ci-checker) package.

**Comment by ChrisWeissmann** [comment]
Currently not all packages ar compatible with our license model. How do we resolve this?
The license checker spits out following packages incompatible:

❌	The License ("LIL") of @nuxt/devalue@1.2.2 is incompatible.
❌	The License ("SEE LICENSE IN LICENSE.md") of abab@2.0.0 is incompatible.
❌	The License ("SEE LICENSE IN LICENSE") of create-react-context@0.2.3 is incompatible.
❌	The License ("BSD-like") of css-select@1.2.0 is incompatible.
❌	The License ("BSD") of doctrine@1.5.0 is incompatible.
❌	The License ("UNKNOWN") of domutils@1.5.1 is incompatible.
❌	The License ("DBAD") of easy-stack@1.0.0 is incompatible.
❌	The License ("BSD") of esutils@2.0.2 is incompatible.
❌	The License ("BSD") of glob-to-regexp@0.3.0 is incompatible.
❌	The License ("DBAD") of js-message@1.0.5 is incompatible.
❌	The License ("DBAD") of js-queue@2.0.0 is incompatible.
❌	The License ("AFLv2.1,BSD") of json-schema@0.2.3 is incompatible.
❌	The License ("Public Domain") of jsonify@0.0.0 is incompatible.
❌	The License ("MIT/X11") of mkdirp@0.3.0 is incompatible.
❌	The License ("DBAD") of node-ipc@9.1.1 is incompatible.
❌	The License ("MIT/X11") of optimist@0.6.1 is incompatible.
❌	The License ("UNKNOWN") of process@0.5.2 is incompatible.
❌	The License ("BSD") of regjsparser@0.1.5 is incompatible.
❌	The License ("BSD") of sax@0.5.8 is incompatible.
❌	The License ("UNKNOWN") of schulcloud-client@0.12.0 is incompatible.
❌	The License ("BSD") of source-map@0.1.43 is incompatible.
❌	The License ("SEE LICENSE IN LICENSE") of video-player@2.4.0 is incompatible.
❌	The License ("UNKNOWN") of vue-toasted@1.1.26 is incompatible.
❌	The License ("MIT/X11") of wordwrap@0.0.2 is incompatible.

**Comment by adrianjost** [comment]
@ChrisWeissmann This may be an issue with my compatiblity check. But personaly I am only getting one issue when running `yarn license-check` and that's from our own video player:

```bash
❌       The License ("SEE LICENSE IN LICENSE") of video-player@2.4.0 is incompatible.
```

Normaly we shouln't allow scripts that are compatible with our license and look for alternatives.

**Comment by adrianjost** [comment]
I added a whitelist option to the module and whitelistet the video-player.

File: package.json [hunk 1]
```diff
@@ -26,15 +26,16 @@
 		"lint:fix:all:stylelint": "yarn lint:all:stylelint --fix",
 		"lint:fix:all:markdownlint": "yarn lint:all:markdownlint",
 		"lint:fix:all:prettier": "yarn lint:all:prettier --write",
-		"lint:ci": "run-p lint:all:*",
+		"lint:ci": "run-s lint:all:*",
```
Comment (adrianjost) [inline package.json]:
```suggestion
		"lint:ci": "run-p lint:all:*",
```

**Comment by ChrisWeissmann** [comment]
If i run it with the command stated in the docs (```license-ci-checker --production```) I still get a long list of violations.

When I run it with the command stated in your comment (```license-check```) everything is fine.

**Comment by adrianjost** [comment]
Oh, yes I forgot to update the docs. Sorry for that

**Comment by adrianjost** [comment]
Could someone else with a mac os device test if `yarn ci` or `yarn license-check`  works? On Windows everything works fine for me. On Linux (Travis), everything works too, but those commands are failing on the machine of @ChrisWeissmann (macOS). 

@janrenz @dominikbruechner could you take a look on it?

**Comment by domibr** [comment]
> Could someone else with a mac os device test if `yarn ci` or `yarn license-check` works? On Windows everything works fine for me. On Linux (Travis), everything works too, but those commands are failing on the machine of @ChrisWeissmann (macOS).
> 
> @janrenz @dominikbruechner could you take a look on it?

Number of entries found: 18

Your package license: AGPL-3.0

Checking dependency licenses for compatibility...
❌	The License ("SEE LICENSE IN LICENSE") of video-player@2.4.0 is incompatible.
error Command failed with exit code 1.

Seems fine! I just get the one incompatible license

**Comment by adrianjost** [comment]
@domibr Hmm that doesn't sound right. @ChrisWeissmann had the same issue on his mac, but I can't reproduce it neighter on windows nor on linux.

**What do you think:** merge it and keep it running on travis as normal and ignore the fact that it's throwing errors on mac os or close this pull for now?

**Comment by adrianjost** [comment]
merge: 👍 
close: 👎

**Comment by ChrisWeissmann** [comment]
I think it's a good tool. The check is only run on Travis correct? It isn't run in the git staging process where all the linting happens. 
If it works on travis I would be fine with keeping it. 
Just wondering if what we have diffrently. What node version do you have?  I am running v11.13.0

**Comment by adrianjost** [comment]
It's running fine on travis and on my local windows machine with Node 8.12.0 and 10.15.0. And yes, it is not processed during lint-staged.

**Comment by domibr** [comment]
I'd say if it runs on travis fine, we can proceed. And research on the mac
problems at some point

**Comment by ChrisWeissmann** [comment]
Just to make sure it's not the node version I tried it with node 10.15.3 
Same result :(
I vote we still run it and try to debug to tool on a mac sometime soon

**Comment by CeEv** [comment]
It show like a nice tool. If it is merged and work well we should look, to use it in server ms too. But should run only, if package.json is changed.

**Comment by ChrisWeissmann** [comment]
I just checked this out again and merged the current master that now has the current client as a submodule. I am recieving errors for 3 packages:
❌	The License ("(OFL-1.1 AND MIT)") of font-awesome@4.7.0 is incompatible.
❌	The License ("UNKNOWN") of nexboard-api-js@1.1.2 is incompatible.
❌	The License ("UNKNOWN") of path.join@1.0.0 is incompatible.

This brings us to the question should submodules be included in the check? I think yes.

Not sure where this would be apropriate to post if not here.

**Comment by adrianjost** [comment]
> I just checked this out again and merged the current master that now has the current client as a submodule. I am recieving errors for 3 packages:
> ❌	The License ("(OFL-1.1 AND MIT)") of font-awesome@4.7.0 is incompatible.
> ❌	The License ("UNKNOWN") of nexboard-api-js@1.1.2 is incompatible.
> ❌	The License ("UNKNOWN") of path.join@1.0.0 is incompatible.
> 
> This brings us to the question should submodules be included in the check? I think yes.
> 
> Not sure where this would be apropriate to post if not here.

The Flag --onlyDirectDependencies says, that submodules should get ignored. But for any reason, it still seems to not work on macos at all :(

**Pull request closed without merging**